Libmesh update

roystgnr opened the pull request:
Summary of changes:

- Exodus IGA fix for multiblock files
- fix vector-index out of range
- Update TIMPI submodule
- Use more macros from autoconf submodule
- Dual shape function modification
- IGA files need RATIONAL\_BERNSTEIN mapping even for w=1.0
- Fix Apple M1 JIT
- Initialize version\_number in Xdr class.
- RB update: Enable EIM on element sides
- Add SVD in preconditioner type
- Only write Xdr header info on proc 0
- Calculator app improvements

Refs #0

Refs #18768 - the bugfix here gets multi-subdomain Exodus IGA files working

This also cherry-picks the previous updates necessary for #19349 - I'm hoping that whatever's causing problems there is in the *non*-libMesh-update commits.

lindsayad committed:
Update libmesh

And enable parallel metaphysicl

Summary of libmesh changes:
  - fix vector-index out of range
  - Update TIMPI submodule
  - Use more macros from autoconf submodule
  - Dual shape function modification
  - IGA files need RATIONAL_BERNSTEIN mapping even for w=1.0
  - Fix Apple M1 JIT
  - Initialize version_number in Xdr class.
  - RB update: Enable EIM on element sides
  - Add SVD in preconditioner type
  - Only write Xdr header info on proc 0
  - Calculator app improvements

Refs #000

lindsayad committed:
Add newsletter update

roystgnr committed:
Further libMesh update

This gets the multi-subdomain IGA fixes I just merged

roystgnr committed:
LaTeX fixes and new entry in libMesh news

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit c3a9eae9c19b69f9e96df2d04036e22209218f89
+Subproject commit 62774ade899b871fb8acde2cc9373a34d3eb12f9
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

loganharbour commented on modules/doc/content/newsletter/2021_11.md:
```diff
@@ -43,4 +43,17 @@ If any issues with the new packages are experienced, please open a new
 
 ## libMesh-level Changes
 
+- Exodus IGA fix for multiblock files
```
We really don't want this to be a boring changelog...

roystgnr replied:
Version 2 is pushed.  It's at least less terse and less mechanical sounding; if it's not sufficiently less boring then in Version 3 I can put more dumb jokes into the commit rather than just the commit message.

loganharbour commented on conda/libmesh/build.sh:
```diff
@@ -75,6 +75,7 @@ fi
 source $SRC_DIR/configure_libmesh.sh
 export INSTALL_BINARY="${SRC_DIR}/build-aux/install-sh -C"
 METHODS="opt oprof devel dbg" configure_libmesh --prefix=${PREFIX}/libmesh \
+                                                --with-future-timpi-dir=${PREFIX}/libmesh \
```
Can you not unify this in configure_libmesh.sh?

roystgnr replied:
I'll wait for @lindsayad to chime in, since I stole this commit from him, but yeah, this looks suspicious.  Would making this change only in conda/libmesh/build.sh mean that we do the build right in conda but then the first person to `update_and_rebuild_libmesh.sh` manually discovers that MetaPhysicL communication isn't compiling?

lindsayad replied:
I also made this change in the update script. The thing is that the install directory is set using different environment variables here and in the update script. I haven’t looked to see whether this build.sh could use LIBMESH_DIR like the update script uses and which our manual users rely on. If so then this could be moved into configure.sh. Anyway it was easiest for me just to follow the prefix configure option

loganharbour replied:
The whole point of that configure script is to make sure that we keep consistent configure flags between conda and user builds; I would prefer that we modify it to support this if possible

loganharbour replied:
Let's actually just require a `PREFIX` var in `configure_libmesh.sh` and add:
```
--prefix=${PREFIX} --with-future-timpi-dir=${PREFIX/libmesh}
```

to the configure flags with 

```
PREFIX=${PREFIX}/libmesh METHODS="opt oprof devel dbg" configure_libmesh
```

in `conda/libmesh/build.sh`

and

```
SRC_DIR=${SCRIPT_DIR}/../libmesh PREFIX=${LIBMESH_DIR} configure_libmesh ...
```

in `update_and_rebuild_libmesh.sh`?

roystgnr replied:
Ah, I see it now.  So this should work as-is.

As far as these two scripts are concerned, it ought to be fine to add `$LIBMESH_DIR` as yet another "scream and die if this env variable isn't defined" entry at the top of configure_libmesh.sh.  I'll do that now and see if anything in CI here screams.  Are there likely to be any users of configure_libmesh.sh *other than* these two scripts, though?

loganharbour replied:
No. It's meant to be used internally. I would probably be okay with removing `SRC_DIR` for `LIBMESH_DIR` with a scream-and-die. I have faith in you making a reasonable change that works cleanly for both install methods

roystgnr replied:
@loganharbour - Github didn't update fast enough for me, so I missed your last comment here before making + pushing my commit.  Want to check that out and see what you think?

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/885658/) on b181e3e wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/19370/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/885668/) on b181e3e wanted to post the following:


## Modules coverage

No coverage report(s) were found for the base commit c5bb2aac28.

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/19370/coverage/xfem)

</details>



This comment will be updated on new commits.

roystgnr committed:
Make libMesh changelog less boring

And Logan, the reviews are pouring in!

Richard Roeper: "Compelling and mesmerizing.  A tour de force!"

Roger Ebert: "I laughed.  I cried.  Those are probably bad things when
reading a technical report."

Gene Siskel: "Don't you put words in my mouth too.  Just because I can't
file a libel suit doesn't mean my estate can't."

roystgnr committed:
Use $LIBMESH_DIR in configure_libmesh.sh

loganharbour commented on conda/libmesh/build.sh:
```diff
@@ -74,8 +74,8 @@ fi
 export INSTALL_BINARY="${SRC_DIR}/build-aux/install-sh -C"
-METHODS="opt oprof devel dbg" configure_libmesh --prefix=${PREFIX}/libmesh \
-                                                --with-vtk-lib=${BUILD_PREFIX}/libmesh-vtk/lib \
+export LIBMESH_DIR=${PREFIX}/libmesh
+METHODS="opt oprof devel dbg" configure_libmesh --with-vtk-lib=${BUILD_PREFIX}/libmesh-vtk/lib \
```
```suggestion
LIBMESH_DIR=${PREFIX}/libmesh METHODS="opt oprof devel dbg" configure_libmesh --with-vtk-lib=${BUILD_PREFIX}/libmesh-vtk/lib \
```

I would say do this; try to export as little as possible with conda

loganharbour reviewed (approved):
We can get this in once we clean up next

roystgnr commented:
Great, thanks.  I have some more TIMPI and libMesh PRs in the pipeline but nothing that needs to go into MOOSE in a big hurry.  If you're keeping track of the next cleanup, go ahead and hit merge on this whenever it's ready?
